go-controller/pkg/ovn/service.go

```diff
@@ -1,10 +1,137 @@
 package ovn
 
 import (
+	"fmt"
 	"github.com/Sirupsen/logrus"
 	kapi "k8s.io/client-go/pkg/api/v1"
+	"net"
 )
 
+func (ovn *Controller) syncServices(services []interface{}) {
+	// For all clusterIP in k8s, we will populate the below slice with
+	// IP:port. In OVN's database those are the keys. We need to
+	// have separate slice for TCP and UDP load-balancers (hence the dict).
+	clusterServices := make(map[string][]string)
+
+	// For all nodePorts in k8s, we will populate the below slice with
+	// nodePort. In OVN's database, nodeIP:nodePort is the key.
+	// We have separate slice for TCP and UDP nodePort load-balancers.
+	// We will get nodeIP separately later.
+	nodeportServices := make(map[string][]string)
+
+	// Go through the k8s services and populate 'clusterServices' and
+	// 'nodeportServices'
+	for _, serviceInterface := range services {
+		service, ok := serviceInterface.(*kapi.Service)
+		if !ok {
+			logrus.Errorf("Spurious object in syncServices: %v",
+				serviceInterface)
+			continue
+		}
+
+		if service.Spec.Type != kapi.ServiceTypeClusterIP &&
+			service.Spec.Type != kapi.ServiceTypeNodePort {
+			continue
+		}
+
+		if service.Spec.ClusterIP == "" {
+			continue
+		}
+
+		for _, svcPort := range service.Spec.Ports {
+			protocol := svcPort.Protocol
+			if protocol == "" || (protocol != TCP && protocol != UDP) {
+				protocol = TCP
+			}
+
+			if service.Spec.Type == kapi.ServiceTypeNodePort {
+				port := fmt.Sprintf("%d", svcPort.NodePort)
+				if protocol == TCP {
+					nodeportServices[TCP] = append(nodeportServices[TCP], port)
+				} else {
+					nodeportServices[UDP] = append(nodeportServices[UDP], port)
+				}
+			}
+
+			if svcPort.Port == 0 {
+				continue
+			}
+
+			key := fmt.Sprintf("%s:%d", service.Spec.ClusterIP, svcPort.Port)
+			if protocol == TCP {
+				clusterServices[TCP] = append(clusterServices[TCP], key)
+			} else {
+				clusterServices[UDP] = append(clusterServices[UDP], key)
+			}
+		}
+	}
+
+	// Get OVN's current cluster load-balancer VIPs and delete them if they
+	// are stale.
+	for _, protocol := range []string{TCP, UDP} {
+		loadBalancer := ovn.getLoadBalancer(kapi.Protocol(protocol))
+		loadBalancerVIPS, err := ovn.getLoadBalancerVIPS(loadBalancer)
+		if err != nil {
+			logrus.Errorf("failed to get load-balancer vips for %s (%v)",
+				loadBalancer, err)
+			continue
+		}
+		if loadBalancerVIPS == nil {
+			continue
+		}
+
+		for vip := range loadBalancerVIPS {
+			if !stringSliceMembership(clusterServices[protocol], vip) {
```

Comment:
An info/debug message about deleting the stale VIP would be useful.

Reply:
I added a debug message.